src/zkp/pifac.rs

```diff
@@ -491,6 +495,84 @@ mod tests {
         Ok(())
     }
 
+    #[test]
+    fn proof_elements_should_be_correct() -> Result<()> {
```

Comment:
todo: This test is a good place to use the `with_random_no_small_factors_proof` approach, since we're starting with a valid input + proof, then modifying the proof one field at at time. That means you would skip the lines defining the input and proof and put everything else in your closure.

nit: please consider breaking up the block of code with some comments or newlines. Maybe separating each type of substitution (e.g. "replace each `BigNumber`; replace each `Commitment`..._) would be a reasonable separation. I just get lost in so many unbroken lines of code.

Reply:
If one field is already wrong, then isn't it redundant to put cumulatively n fields wrong in the nth assert? rather than testing whether each field is wrong separately does not work?